beater/beater.go

```diff
@@ -184,33 +187,56 @@ func (bt *beater) Run(b *beat.Beat) error {
 		return nil
 	}
 
-	go notifyListening(bt.config, pub.Client().Publish)
-
 	bt.mutex.Lock()
 	if bt.stopped {
 		defer bt.mutex.Unlock()
 		return nil
 	}
 
-	bt.server = newServer(bt.config, tracer, pub.Send)
+	bt.server, err = newServer(bt.config, tracer, pub.Send)
+	if err != nil {
+		bt.logger.Error("failed to create new server:", err)
+		return nil
+	}
 	bt.mutex.Unlock()
 
 	var g errgroup.Group
 	g.Go(func() error {
 		return run(bt.server, lis, bt.config)
 	})
+
+	if bt.isServerAvailable(10 * time.Second) {
+		go notifyListening(bt.config, pub.Client().Publish)
+	}
+
 	if traceListener != nil {
 		g.Go(func() error {
 			return bt.server.Serve(traceListener)
 		})
 	}
+
 	if err := g.Wait(); err != http.ErrServerClosed {
 		return err
 	}
 	bt.logger.Infof("Server stopped")
 	return nil
 }
 
+func (bt *beater) isServerAvailable(timeout time.Duration) bool {
+	// following an example from https://golang.org/pkg/net/
+	// dial into tcp connection to ensure listener is ready, send get request and read response,
+	// in case tls is enabled, the server will respond with 400,
+	// as this only checks the server is up and reachable errors can be ignored
+	conn, err := net.DialTimeout("tcp", bt.config.Host, timeout)
+	if err != nil {
+		return false
+	}
+	fmt.Fprintf(conn, "GET / HTTP/1.0\r\n\r\n")
+	bufio.NewReader(conn).ReadString('\n')
```

Comment:
iirc, you said we need to read something on the conn for an error to be thrown at `Close` - if so, can we read just a single byte (`ReadByte()`) instead?  Does apm-server log anything if we only read part of the response?